### PR TITLE
Unified admin grid columns and added XML configuration

### PR DESCRIPTION
**Pull request by sreichel**
### Description (*)

- makes all `entity_id` grid columns searchable `from` `to`
- disallow/ignore non-numeric input on price/currency/range grid columns (*)
- removed useless `"'sortable' => true"`
- bugfix: revmoved `return parent::renderCss() . ' a-right';`

(*) Note: instead of having zero results, as suggested in mentioned issue, it will ignore input and return all products.

### Fixed Issues (if relevant)

1. Fixes OpenMage/magento-lts#3158

## Update

The configuratuion for grid column default values went from php-property to XML config.

To customize see ... `app/code/core/Mage/Adminhtml/etc/config.xml`

```xml
    <grid>
        <column>
            <default>
                <index>
                    <entity_id>
                        <align>right</align>
                        <type>number</type>
                    </entity_id>
                </index>
                <type>
                    <action>
                        <filter>false</filter>
                        <sortable>false</sortable>
                        <width>40</width>
                    </action>
                    <currency>
                        <align>right</align>
                        <index>price</index>
                    </currency>
                    <date>
                        <width>140</width>
                    </date>
                    <datetime>
                        <width>170</width>
                    </datetime>
                    <number>
                        <align>right</align>
                    </number>
                    <price>
                        <align>right</align>
                        <index>price</index>
                    </price>
                    <store>
                        <index>store_id</index>
                        <store_view>true</store_view>
                        <width>160</width>
                    </store>
                </type>
            </default>
        </column>
    </grid>
```

### CODE REVIEW

**Comment by ADDISON74** [comment]
If we fix this problem, all the grids in the Backend will have to be revised. For example, the **Manage Customers** grid, we are allowed to input strings on the ID column for the From and To inputs.

**Comment by sreichel** [comment]
> If we fix this problem, all the grids in the Backend will have to be revised

Yes, i know, but please not here in this PR.

Better make small steps ...

**Comment by ADDISON74** [comment]
@sreichel - This is the grids list which have columns that allow searching strings and return zero as results. In order to check I had to change records in the database. I hope I did not miss anything. This PR solves only one of the issues **Catalog > Products - Price column**.

**Sales > Orders**
- [ ] G.T. (Base) column
- [ ] G.T. (Purchased) column

**Sales > Invoices**
- [ ] Amount column

**Sales > Shipments**
- [ ] Total Qty column

**Sales > Credit Memos**
- [ ] Refunded column

**Sales > Tax > Manage Tax Zone & Rates**
- [ ] Rate column

**Catalog > Products**
- [ ] Price column

**Catalog > Products > Edit Product > Related Products tab**
- [ ] Price column
- [ ] Position colum

**Catalog > Products > Edit Product > Up-sells Products tab**
- [ ] Price column
- [ ] Position colum

**Catalog > Products > Edit Product > Cross-sells Products tab**
- [ ] Price column
- [ ] Position colum

**Catalog > Products > Edit Product > Product Tags tab**
- [ ] # of Use column

**Catalog > Products > Edit Product > Associated Products tab (configurable)**
- [ ] Price column

**Catalog > Categories > Category Products tab**
- [ ] Price column
- [ ] Position column

**Catalog > Search Terms**
- [ ] Results column
- [ ] Number of Uses column

**Promotions > Shopping Cart Price Rules**
- [ ] Times used column

**Newsletter > Newsletter Queue**
- [ ] Processed column
- [ ] Recipients column

**Newsletter > Newsletter Problem Reports**
- [ ] Error Code column

**Reports > Shopping Cart > Abandoned carts**
- [ ] Number of Items column
- [ ] Quantity of Items column
- [ ] Subtotal column

**Reports > Producs > Low stock**
- [ ] Stock Qty column

**Reports > Search Terms**
- [ ] Results column
- [ ] Hits column

**Review by ADDISON74 — APPROVED** [review]
Nice work. I checked all the files and the result in the Backend.

**Comment by ADDISON74** [comment]
I have identified a problem. In **Catalog > Manage Products** on the ID column it is now possible to search within a range. Similar grids are in Related Products, Up-sells, Cross-sells, Associate Products and many more, but which only remained with the search for an ID. Shouldn't they also be modified to be unified?

**Comment by sreichel** [comment]
Not all grids are covered yet (having another index then entity_id)

If you find some, please list them.

**Comment by ADDISON74** [comment]
Grids with ID columns which should have From/To inputs.

- [ ] Catalog > Manage Products > Product Edit > Related Products

- [ ] Catalog > Manage Products > Product Edit > Up-sells

- [ ] Catalog > Manage Products > Product Edit > Cross-sells

- [ ] Catalog > Manage Products > Product Edit > Product Reviews

- [ ] Catalog > Manage Products > Product Edit > Associated Products (configurable)

- [ ] Catalog > Manage Categories > Category Products tab

- [ ] Catalog > URL Rewrite Management

- [ ] Catalog > Search Terms

- [ ] Catalog > Review and Ratings > Customer Reviews > Pending Reviews

- [ ] Catalog > Review and Ratings > Customer Reviews > All Reviews

- [ ] Catalog > Google Sitemap

- [ ] Promotions > Catalog Price Rules

- [ ] Promotions > Shopping Cart Price Rules

- [ ] Newsletter > Newsletter Templates

- [ ] Newsletter > Newsletter Queue

- [ ] Newsletter > Newsletter Subscribers

- [ ] Newsletter > Newsletter Problems Reports

- [ ] System > Webservices > SOAP Users

- [ ] System > Webservices > SOAP Roles

- [ ] System > Webservices > Rest Roles

- [ ] System > Webservices > OAuth Consumers

- [ ] System > Webservices > OAuth Authorized Tokens

- [ ] System > Webservices > My Applications

- [ ] System > Import/Export > Dataflow Profiles

- [ ] System > Import/Export > Advanced Profiles

- [ ] System > Transactional Emails

- [ ] System > Permissions > Users

- [ ] System > Permissions > Roles

- [ ] System > Permissions > Variables

- [ ] System > Permissions > Blocks

Before moving on to the changes, I think we should evaluate these grids, because not all of them need changes, but those where the need to search within a rage is really helpful.

**Comment by sreichel** [comment]
Thanks. I'll do it step by step :)

**Comment by sreichel** [comment]
Please checkout latest commit.

This was fixed in  [app/code/core/Mage/Adminhtml/Block/Widget/Grid.php](https://github.com/OpenMage/magento-lts/pull/4225/commits/aec46253c6ce674fd5bfcf2a56ebb18017574762#diff-1d4d6a90dffd32740adc95fb72207082f6ccf85705956c397976b86e9d7ec98b)

**Comment by sreichel** [comment]
I have checked that ones hat should be already fixed.

Leave everything else as it is?

**Comment by ADDISON74** [comment]
The grids you selected for changing the ID column are fine. Do you have any other ideas related to this PR?

**Comment by sreichel** [comment]
Last changes ...

- unified storeview select columns
- use config values from exml to customization easy
- added test

File: tests/unit/Mage/Adminhtml/Block/Widget/GridTest.php [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    /**
+     * @dataProvider provideAddColumnDefaultData
+     * @group Mage_Adminhtml
+     * @group Mage_Adminhtml_Block
+     */
+    public function testAddColumnDefaultData(array $expectedResult, array $column): void
+    {
+        $node = Mage::getConfig()->getNode('grid_columns');
+        $this->assertSame($expectedResult, $this->subject->addColumnDefaultData($column));
```
Comment (sreichel) [inline tests/unit/Mage/Adminhtml/Block/Widget/GridTest.php:44]:
```suggestion
        $this->assertSame($expectedResult, $this->subject->addColumnDefaultData($column));
```

**Comment by sreichel** [comment]
Done.

**Comment by ADDISON74** [comment]
from_to

The change for the grid columns that have **From/To** texts is extremely useful. By default I would leave **From/To** on the ID column in front of the input boxes. If someone wants to save space just use empty strings in /app/design/frontend/default/{theme_name}/locale/en_US/translate.csv

```csv
"From",""
"To",""
```

We discussed about the placeholder in the past in a PR. Using it now solves the situation because the text is now inside the input box. The From and To texts are no longer needed, but this must be left to everyone's discretion.

**Comment by ADDISON74** [comment]
In **Catalog > Manage Products**, the column **Price** still has From and To in front of the input boxes. So far it is the only grid having this. Please see the screenshot bellow

from_to

**Comment by ADDISON74** [comment]
If we remove From/To we should do it for date boxes too.

**Comment by ADDISON74** [comment]
These input boxes could create confusion. Here is a scenario From=900 To=100, there are products in this reverse range but no results will be displayed.

**Comment by ADDISON74** [comment]
If someone wants to hide **From, To, In** labels to save space by editing the translate.csv file, then he must change the boxes.css file as well

**from**

```css
.grid tr.filter .range input {
    width: 60px !important;
    margin-top: 0;
}

.grid tr.filter .range select {
    float: right;
    width: 65px !important;
    margin-top: 0;
}
```

**to**

```css
.grid tr.filter .range input {
    width: 95px !important;
    margin-top: 0;
}

.grid tr.filter .range select {
    margin-top: 0;
}
```

Changing **From, To, In** creates an issue with the placeholder, it won't appear anymore. My opinion is to get rid off the From, To, In labels.

**Comment by sreichel** [comment]
- changing CSV to empty string would change all occurrences
- have re-added "from" "to" in legacy theme
- leaving it to date and price ranges was intended - as it ever was there and make sense to me.
- there are the new CSS-selectors (`.filter-range`, `.filter-price`) that can be used to show/hide labels

**Comment by ADDISON74** [comment]
It is pretty strange to see two times the same text, From From, To To

from

My opinion is that if we start using the placeholder then we have to give up the labels.

**Comment by sreichel** [comment]
Date-placeholder should have "From Date", or "dd/mm/yy" to make it perfect  ...

PR welcome.

**Comment by sreichel** [comment]
@ADDISON74 thanks for your reviews. If you have improvements, please use the "suggestion"-feature or make a PR to my repo.

**Comment by sreichel** [comment]
> It is pretty strange to see two times the same text, From From, To To

Made placeholders invisible in legacy theme.

File: app/code/core/Mage/Oauth/Block/Adminhtml/Oauth/Consumer/Grid.php [hunk 2]
```diff
@@ -66,7 +66,7 @@ protected function _prepareCollection()
     protected function _prepareColumns()
     {
         $this->addColumn('entity_id', [
-            'header' => Mage::helper('oauth')->__('ID'), 'index' => 'entity_id', 'align' => 'right', 'width' => '50px'
+            'header' => Mage::helper('oauth')->__('ID'), 'index' => 'entity_id'
```
Comment (kiatng) [inline app/code/core/Mage/Oauth/Block/Adminhtml/Oauth/Consumer/Grid.php:69]:
```suggestion
            'header' => Mage::helper('oauth')->__('ID'),
            'index' => 'entity_id'
```

File: app/code/core/Mage/Adminhtml/Block/Widget/Grid.php [hunk 3]
```diff
@@ -349,6 +340,61 @@ public function addColumn($columnId, $column)
...
+        return $column;
+    }
+
+    public function getConfigDefaultColumnSettings(): array
+    {
+        if (is_null($this->defaultColumnSettings)) {
+            $configNode = Mage::getConfig()->getNode('grid/column/default');
+            # should be called only once
+            if ($configNode === false) {
+                Mage::app()->getCacheInstance()->cleanType('config');
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Widget/Grid.php:383]:
```suggestion
                Mage::app()->getCacheInstance()->cleanType('config');
                $configNode = Mage::getConfig()->getNode('grid/column/default');
```

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Widget/Grid.php:383]: Clean cache is only required when you add this PR, but have config-cache enabled. In this case `getNode('grid/column/default')` will return false.

**Comment by ADDISON74** [comment]
> > It is pretty strange to see two times the same text, From From, To To
> 
> Made placeholders invisible in legacy theme.

It is much better.

**Comment by kiatng** [comment]
@sreichel Conflict to resolve.

File: app/code/core/Mage/Adminhtml/Block/Widget/Grid.php (hunk 3 above)
Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Block/Widget/Grid.php:383]: Line 385: 
```php
$config = $configNode->asArray(); // $configNode can be false if not for assignment in line 384 
```

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Widget/Grid.php:383]: Oh. Get the "Suggested change" wrong ... :)

**Review by ADDISON74 — APPROVED** [review]

**Review by kiatng — APPROVED** [review]

**Pull request merged**